### PR TITLE
[WIP] add the SparseGPT Pruner Backbone

### PR DESCRIPTION
**Pull request by Currycurrycurry**
## Type of Change

feature

## Description

SparseGPT 

## Expected Behavior & Potential Risk

Still working on it

## How has this PR been tested?

by shell script

## Dependency Change?

No

### CODE REVIEW

**Comment by Currycurrycurry** [comment]
23/05/24 Update: Construct the basic code structure without debugging, will further complete the details and improve the overall readability. @wenhuach21

File: neural_compressor/compression/pruner/pruners.py [hunk 1]
```diff
@@ -901,3 +904,251 @@ def print_progressive_sparsity(self):
...
+            inp = inp.t()
+        self.H *= self.nsamples / (self.nsamples + tmp)
+        self.nsamples += tmp
+        inp = math.sqrt(2 / self.nsamples) * inp.float()
+        self.H += inp.matmul(inp.t()) # XX⊤的计算
+
+    def fasterprune(
+        self, sparsity, prunen=0, prunem=0, blocksize=128, percdamp=.01
+    ):
+        ############1. 获得Hinv（inverse Hessian Matrix）############################
```
Comment (chensuyue) [inline neural_compressor/compression/pruner/pruners.py:1063]: Please review those annotation, or translate into English.

Comment (Currycurrycurry) [reply neural_compressor/compression/pruner/pruners.py]: done

File: neural_compressor/compression/pruner/pruners.py [hunk 2]
```diff
@@ -901,3 +904,255 @@ def print_progressive_sparsity(self):
...
+        # get sample inputs
+        inputs = torch.zeros(
+        (self.config['nsamples'], model.seqlen, model.config.hidden_size), device=dev)
+
+        # get sample outputs
+        outs = torch.zeros_like(inputs)
+
+        # get cache for attention mask input 
+        cache = {'i': 0, 'attention_mask': None}
+        class Catcher(nn.Module):
```
Comment (wenhuach21) [inline neural_compressor/compression/pruner/pruners.py:974]: @WeiweiZhang1 please help refine the code

File: neural_compressor/compression/pruner/pruners.py [hunk 3]
```diff
@@ -901,3 +904,255 @@ def print_progressive_sparsity(self):
...
+    def fasterprune_masks(self, dataloader, model):
+        ################## preparing for pruning ##################
+        # get device
+        dev = model.device
+        
+        # get layers
+        layers = model.model.decoder.layers
+
+        model.seqlen = model.config.max_position_embeddings
+
```
[the remark this answers is not in the document]
Comment (Currycurrycurry) [reply neural_compressor/compression/pruner/pruners.py:964]: ok, the logic is correct without bugs till now, but it may need further reconstruction for adaptation to our framework.

[the remark this answers is not in the document]
Comment (Currycurrycurry) [reply neural_compressor/compression/pruner/pruners.py:964]: since sparsegpt is a layer-wise post-training method without the need of gradients like the retraining-free pruner, I put all the logic to the on_train_begin method of the pruner class, which may need further splitting to a new pattern class and criterion class (need discussion with @WeiweiZhang1 )

File: neural_compressor/compression/pruner/pruners.py [hunk 4]
```diff
@@ -901,3 +904,264 @@ def print_progressive_sparsity(self):
...
+        return res
+
+
+    def fasterprune_masks(self, dataloader, model):
+        ################## preparing for pruning ##################
+        # get device
+        dev = model.device
+        
+        # get layers
+        layers = model.model.decoder.layers
```
Comment (wenhuach21) [inline neural_compressor/compression/pruner/pruners.py:961]: one key issue you have not addressed. This code only works for some models.

Comment (WeiweiZhang1) [reply neural_compressor/compression/pruner/pruners.py:961]: I will solve it
